src/main/java/com/google/api/codegen/CodeGeneratorApi.java

```diff
@@ -123,27 +121,26 @@ protected void process() throws Exception {
       String factory = generator.getFactory();
       String id = generator.getId();
 
-      GapicProviderFactory providerFactory = createProviderFactory(model, factory);
+      GapicProviderFactory<GapicProvider<? extends Object>> providerFactory = createProviderFactory(
+          model, factory);
       List<GapicProvider<? extends Object>> providers = providerFactory.create(model, apiConfig, id);
       for (GapicProvider<? extends Object> provider : providers) {
         provider.generate(options.get(OUTPUT_FILE));
       }
     }
   }
 
-  private static GapicProviderFactory createProviderFactory(final Model model, String factory) {
+  private static GapicProviderFactory<GapicProvider<? extends Object>> createProviderFactory(
+      final Model model, String factory) {
     @SuppressWarnings("unchecked")
-    GapicProviderFactory provider = GeneratorBuilderUtil.createClass(
-        factory,
-        GapicProviderFactory.class,
-        new Class<?>[] {},
-        new Object[] {},
-        "generator", new GeneratorBuilderUtil.ErrorReporter() {
-        @Override
-        public void error(String message, Object... args) {
-          model.addDiag(Diag.error(SimpleLocation.TOPLEVEL, message, args));
-        }
-      });
+    GapicProviderFactory<GapicProvider<? extends Object>> provider = GeneratorBuilderUtil
```

Comment:
I see. Good point. I guess it's best to wait until the Discovery case is added before thinking about any additional abstraction.